bluepysnap/edges.py

```diff
@@ -480,7 +482,37 @@ def _get(self, selection, properties=None):
 
         return result
 
-    def ids(self, group=None, limit=None, sample=None):
+    def _edge_ids_by_filter(self, queries, raise_missing_prop):
+        """Return edge IDs if their properties match the `queries` dict.
+
+        `props` values could be:
+            pairs (range match for floating dtype fields)
+            scalar or iterables (exact or "one of" match for other fields)
+
+        You can use the special operators '$or' and '$and' also to combine different queries
+        together.
+
+        Examples:
+            >>> self._edge_ids_by_filter({ Edge.POST_SECTION_ID: (0, 1),
+            >>>                            Edge.AXONAL_DELAY: (.5, 2.) })
+            >>> self._edge_ids_by_filter({'$or': [{ Edge.PRE_X_CENTER: [2, 3]},
+            >>>                              { Edge.POST_SECTION_POS: (0, 1),
+            >>>                              Edge.SYN_WEIGHT: (0.,1.4) }]})
+
+        """
+        properties = query.get_properties(queries)
+        unknown_props = properties - self.property_names
+        if raise_missing_prop and unknown_props:
+            raise BluepySnapError(f"Unknown edge properties: {unknown_props}")
+        res = []
+        ids = self.ids(None)
+        chunk_size = int(1e8)
+        for chunk in np.array_split(ids, 1 + len(ids) // chunk_size):
+            data = self.get(chunk, properties - unknown_props)
+            res.extend(chunk[query.resolve_ids(data, self.name, queries)])
+        return res
```

Comment:
We should use numpy array with IDS_DTYPE instead. I wrote the snippet with a list but I forgot we return arrays ...

Reply:
btw wdyt about
```
        res = []
        ids = self.ids(None)
        chunk_size = int(1e8)
        for chunk in np.array_split(ids, 1 + len(ids) // chunk_size):
            data = self.get(chunk, properties - unknown_props)
            res.append(chunk[query.resolve_ids(data, self.name, queries)])
        return np.concatenate(res, dtype=IDS_DTYPE)
```

Reply:
This needs to be benchmarked. I did this housand times but keep forgetting about it. What s the best between what you did and concatenate at every chunk :
```
        res = np.empty(dtype=IDS_DTYPE)
        ids = self.ids(None)
        chunk_size = int(1e8)
        for chunk in np.array_split(ids, 1 + len(ids) // chunk_size):
            data = self.get(chunk, properties - unknown_props)
            res.concatenate([res, chunk[query.resolve_ids(data, self.name, queries)]])
        return res
```
(or something equivalent I did not run this piece of code just writing it live) ?

Reply:
Did you manage to contain the memory consumption with the chunks?

Reply:
> Did you manage to contain the memory consumption with the chunks?

Well, it requires less memory for sure. 1.6Tb edges can be processed with `--mem=200g` and chunked version. The non-chunked version can process only with `--mem=0` which is around 500g depending on the machine.

Reply:
> This needs to be benchmarked.

Not much difference in terms of memory. Used the below benchmark
```
pip install memory_profiler
```

```python
from memory_profiler import profile
import numpy as np


np.random.seed(666)

@profile
def benchmark_append():
    res = []
    for i in range(10):
        start = np.random.randint(-10000, 10000)
        end = np.random.randint(100000, 1000000)
        res.append([i for i in range(start, end)])
    return np.concatenate(res)

@profile
def benchmark_extend():
    res = []
    for i in range(10):
        start = np.random.randint(-10000, 10000)
        end = np.random.randint(100000, 1000000)
        res.extend([i for i in range(start, end)])
    return np.array(res)

if __name__ == '__main__':
    benchmark_append()
    benchmark_extend()
```